File: app/src/game.ts
// TypeScript port of gomoku/game.py.
// 9x9 free-style gomoku: first to 5-in-a-row wins. State is canonicalized so
// plane 0 is always the side-to-move's stones; plane 1 is opponent stones.
// `apply()` flips perspective and snapshots the pre-move canonical board onto
// the head of `history` (most-recent first; up to HISTORY_PLY-1 entries) —
// mirrors gomoku/state_ops.py:apply_move_arrays.
// `toPlanes(n)` supports two layouts:
//   n=3:  [my, opp, const-1]               — legacy single-frame (epoch-100 model)
//   n=17: [my×H, opp×H, const-1]  H=8     — AlphaZero-style history (WL series)

export const BOARD_SIZE = 9;
export const N_ACTIONS = BOARD_SIZE * BOARD_SIZE; // 81
export const WIN_LEN = 5;
export const HISTORY_PLY = 8; // matches gomoku/game.py HISTORY_PLY

const PLANE = BOARD_SIZE * BOARD_SIZE; // 81

const DIRS: ReadonlyArray<readonly [number, number]> = [
  [0, 1],
  [1, 0],
  [1, 1],
  [1, -1],
];

/** Canonical GameState. board is length 162 = 2 planes * 81. */
export class GameState {
  // Plane 0 (offsets 0..80)   = side-to-move's stones.
  // Plane 1 (offsets 81..161) = opponent's stones.
  readonly board: Uint8Array;
  readonly moveCount: number;
  // Up to HISTORY_PLY-1 (=7) past canonical boards, most-recent first.
  // Each entry is the pre-flip canonical board (length 2*PLANE) as observed
  // by the mover at that ply.
  readonly history: ReadonlyArray<Uint8Array>;

  constructor(
    board: Uint8Array,
    moveCount: number,
    history: ReadonlyArray<Uint8Array> = [],
  ) {
    this.board = board;
    this.moveCount = moveCount;
    this.history = history;
  }

  static initial(): GameState {
    return new GameState(new Uint8Array(2 * PLANE), 0, []);
  }

  /** Length-81 mask of legal (empty) actions. */
  legalMask(): Uint8Array {
    const out = new Uint8Array(N_ACTIONS);
    const b = this.board;
    for (let i = 0; i < PLANE; i++) {
      out[i] = b[i] === 0 && b[i + PLANE] === 0 ? 1 : 0;
    }
    return out;
  }

  legalActions(): number[] {
    const mask = this.legalMask();
    const out: number[] = [];
    for (let i = 0; i < N_ACTIONS; i++) if (mask[i]) out.push(i);
    return out;
  }

  /** Return a NEW state with `action` played by the side-to-move; perspective flipped. */
  apply(action: number): GameState {
    if (action < 0 || action >= N_ACTIONS) {
      throw new Error(`out of range: ${action}`);
    }
    if (this.board[action] || this.board[action + PLANE]) {
      throw new Error(`illegal move ${action} on occupied square`);
    }
    // Snapshot the pre-move canonical board (matches Python's state_ops.apply_move_arrays).
    const snapshot = new Uint8Array(this.board);
    const next = new Uint8Array(2 * PLANE);
    // Swap planes during copy: my (plane 0) becomes opponent (plane 1) for next state.
    for (let i = 0; i < PLANE; i++) {
      next[i] = this.board[i + PLANE];
      next[i + PLANE] = this.board[i];
    }
    next[action + PLANE] = 1;
    const nextHistory = [snapshot, ...this.history.slice(0, HISTORY_PLY - 1)];
    return new GameState(next, this.moveCount + 1, nextHistory);
  }

  /** Returns [done, value_from_side_to_move_perspective].
   *
   * Called AFTER apply(), so plane 1 holds the player who just moved.
   * If they have 5-in-a-row, the current side-to-move just lost (value=-1).
   */
  isTerminal(): { done: boolean; value: number } {
    if (hasFiveInARow(this.board, PLANE)) {
      return { done: true, value: -1.0 };
    }
    if (this.moveCount >= N_ACTIONS) return { done: true, value: 0.0 };
    return { done: false, value: 0.0 };
  }

  /** Return Float32Array of shape [nPlanes, 9, 9] flattened.
   *
   * nPlanes=3:  [my, opp, const-1]                — legacy
   * nPlanes=17: AlphaZero-style 2*HISTORY_PLY history planes + const-1, matching
   *             gomoku/game.py:to_planes layout. Slots beyond available history
   *             stay zero.
   */
  toPlanes(nPlanes: number = 3): Float32Array {
    if (nPlanes === 3) {
      const out = new Float32Array(3 * PLANE);
      for (let i = 0; i < PLANE; i++) {
        out[i] = this.board[i];
        out[i + PLANE] = this.board[i + PLANE];
        out[i + 2 * PLANE] = 1.0;
      }
      return out;
    }
    if (nPlanes === 2 * HISTORY_PLY + 1) {
      const H = HISTORY_PLY;
      const out = new Float32Array(nPlanes * PLANE);
      // Plane 0 (my, t) and plane H (opp, t)
      for (let i = 0; i < PLANE; i++) {
        out[i] = this.board[i];
        out[i + H * PLANE] = this.board[i + PLANE];
      }
      // Planes 1..H-1 / H+1..2H-1: past frames, perspective-corrected.
      // history[k-1] = canonical board observed by mover at t-k. If k is odd
      // the mover was the opposite side, so swap planes when reading.
      for (let k = 1; k < H; k++) {
        const past = this.history[k - 1];
        if (!past) break;
        const myOffset = k % 2 === 0 ? 0 : PLANE;
        const oppOffset = k % 2 === 0 ? PLANE : 0;
        for (let i = 0; i < PLANE; i++) {
          out[k * PLANE + i] = past[myOffset + i];
          out[(H + k) * PLANE + i] = past[oppOffset + i];
        }
      }
      // Last plane (2H): constant 1.0
      for (let i = 0; i < PLANE; i++) out[2 * H * PLANE + i] = 1.0;
      return out;
    }
    throw new Error(`unsupported nPlanes=${nPlanes} (expected 3 or 17)`);
  }
}

/** Check if `plane` (length-81 slice starting at `offset` in `board`) has a 5-in-a-row. */
function hasFiveInARow(board: Uint8Array, offset: number): boolean {
  for (const [dr, dc] of DIRS) {
    for (let r0 = 0; r0 < BOARD_SIZE; r0++) {
      for (let c0 = 0; c0 < BOARD_SIZE; c0++) {
        if (!board[offset + r0 * BOARD_SIZE + c0]) continue;
        const rEnd = r0 + dr * (WIN_LEN - 1);
        const cEnd = c0 + dc * (WIN_LEN - 1);
        if (rEnd < 0 || rEnd >= BOARD_SIZE || cEnd < 0 || cEnd >= BOARD_SIZE) continue;
        let ok = true;
        for (let k = 1; k < WIN_LEN; k++) {
          if (!board[offset + (r0 + dr * k) * BOARD_SIZE + (c0 + dc * k)]) {
            ok = false;
            break;
          }
        }
        if (ok) return true;
      }
    }
  }
  return false;
}

export function actionToStr(action: number): string {
  const r = Math.floor(action / BOARD_SIZE);
  const c = action % BOARD_SIZE;
  return `${String.fromCharCode("a".charCodeAt(0) + c)}${r + 1}`;
}

export function strToAction(s: string): number {
  s = s.trim().toLowerCase();
  if (s.length < 2) throw new Error(`bad move: ${s}`);
  const col = s.charCodeAt(0) - "a".charCodeAt(0);
  const row = parseInt(s.slice(1), 10) - 1;
  if (!Number.isInteger(row) || col < 0 || col >= BOARD_SIZE || row < 0 || row >= BOARD_SIZE) {
    throw new Error(`out of range: ${s}`);
  }
  return row * BOARD_SIZE + col;
}

File: gomoku/game.py
"""Free-style gomoku: first to 5-in-a-row wins, no opening restrictions.

State is stored as two boolean planes (current-player stones, opponent stones).
After every move we flip perspective so the side-to-move is always plane 0.
This canonical form is what gets fed to the network and the MCTS tree.

AlphaZero-style history: the `to_planes()` output includes `HISTORY_PLY` past
canonical board snapshots per side, so the network can see threats forming
across multiple plies — defense is much easier to learn when "opponent just
placed a 3rd stone in a row" is visible directly rather than inferred from a
static snapshot.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from gomoku import state_ops

BOARD_SIZE = 9
N_ACTIONS = BOARD_SIZE * BOARD_SIZE
WIN_LEN = 5
HISTORY_PLY = 8                                  # past plies of each side in the input
N_INPUT_PLANES = 2 * HISTORY_PLY + 1             # 8 me-hist + 8 opp-hist + 1 const = 17

# 4 line directions (the other 4 are their negatives, so 4 covers both ways)
_DIRS = ((0, 1), (1, 0), (1, 1), (1, -1))


@dataclass
class GameState:
    """Canonical: plane 0 = side-to-move's stones, plane 1 = opponent's stones.

    `history` holds up to `HISTORY_PLY - 1` past canonical boards, most-recent
    first. Each entry is a (2, BOARD_SIZE, BOARD_SIZE) bool array that was the
    canonical `board` immediately before that move's apply() flipped perspective.
    """

    board: np.ndarray              # (2, N, N) bool
    move_count: int
    history: tuple = field(default_factory=tuple)  # tuple of (2, N, N) bool arrays

    @classmethod
    def initial(cls) -> "GameState":
        return cls(
            board=np.zeros((2, BOARD_SIZE, BOARD_SIZE), dtype=bool),
            move_count=0,
            history=(),
        )

    def legal_mask(self) -> np.ndarray:
        """Return (N_ACTIONS,) bool — True where empty (legal to place)."""
        return state_ops.legal_mask(self.board)

    def legal_actions(self) -> np.ndarray:
        return np.flatnonzero(self.legal_mask())

    def apply(self, action: int) -> "GameState":
        """Place a stone for the side-to-move at action, then flip perspective.

        The returned state has perspective = the next player to move. The pre-move
        canonical board (`board` BEFORE the stone was placed and BEFORE the flip —
        i.e. what the mover *observed* when deciding) is pushed onto the head of
        `history`. That way each history slot encodes a real past observation
        distinct from the current frame.
        """
        new_board, new_history = state_ops.apply_move_arrays(
            self.board,
            self.history,
            action,
            history_ply=HISTORY_PLY,
        )
        return GameState(
            board=new_board, move_count=self.move_count + 1, history=new_history
        )

    def is_terminal(self) -> tuple[bool, float]:
        """Check for terminal state.

        Returns (done, value_from_side_to_move_perspective).
        value is +1 if side-to-move just won (impossible — see below),
                -1 if opponent just won, 0 for draw, undefined otherwise.

        Convention: this is called AFTER apply(), so plane 1 holds the player who
        just moved. We check plane 1 for a 5-in-a-row.
        """
        return state_ops.terminal_value(self.board, self.move_count)

    def to_planes(self) -> np.ndarray:
        """Return (N_INPUT_PLANES, N, N) float32 input for the network.

        Layout (mirrors AlphaZero Go, scaled down to HISTORY_PLY plies):
          0           : my stones at t   (= side-to-move's stones, current frame)
          1..H-1      : my stones at t-1, t-2, ..., t-(H-1)
          H           : opp stones at t
          H+1..2H-1   : opp stones at t-1, ..., t-(H-1)
          2H          : constant 1.0

        For history step k (1..H-1) the *meaning* of "my" rotates because the
        perspective flipped each ply: at step k the canonical mover was the same
        side as now iff k is even, opposite iff k is odd. So we read the matching
        plane out of `history[k-1]`.
        """
        H = HISTORY_PLY
        out = np.zeros((N_INPUT_PLANES, BOARD_SIZE, BOARD_SIZE), dtype=np.float32)
        out[0] = self.board[0]
        out[H] = self.board[1]
        for k, past_board in enumerate(self.history, start=1):
            if k >= H:
                break
            if k % 2 == 0:
                # Same parity → past mover same side as now → past_board[0] == my color
                out[k] = past_board[0]
                out[H + k] = past_board[1]
            else:
                # Opposite parity → swap
                out[k] = past_board[1]
                out[H + k] = past_board[0]
        out[2 * H] = 1.0
        return out


def _has_five_in_a_row(plane: np.ndarray) -> bool:
    # Kept for existing imports/tests; the implementation now lives behind the
    # state_ops boundary so a native helper can replace it in one place.
    return state_ops.has_five_in_a_row(plane)


# ----- 8-fold symmetry for data augmentation -----
# A square board has 8 symmetries (D4): 4 rotations × 2 reflections.
# For each, the action index (r, c) maps to (r', c') under the same transform.

def _sym_board(board: np.ndarray, sym: int) -> np.ndarray:
    """Apply one of 8 D4 symmetries to a (..., 9, 9) array."""
    rot = sym % 4
    flip = sym // 4
    out = np.rot90(board, rot, axes=(-2, -1))
    if flip:
        out = np.flip(out, axis=-1)
    return np.ascontiguousarray(out)


def _sym_policy(policy: np.ndarray, sym: int) -> np.ndarray:
    """Apply same symmetry to a (81,) policy vector."""
    p = policy.reshape(BOARD_SIZE, BOARD_SIZE)
    return _sym_board(p, sym).reshape(-1)


def augment(planes: np.ndarray, policy: np.ndarray) -> list[tuple[np.ndarray, np.ndarray]]:
    """Return all 8 (planes, policy) pairs under D4 symmetry."""
    return [(_sym_board(planes, s), _sym_policy(policy, s)) for s in range(8)]


def augment_with_aux(
    planes: np.ndarray, policy: np.ndarray, aux_policy: np.ndarray
) -> list[tuple[np.ndarray, np.ndarray, np.ndarray]]:
    """Like `augment`, but carries a SECOND policy vector (the auxiliary
    opponent-reply target) through the SAME D4 symmetry as the position.

    This is the load-bearing alignment guarantee for the aux head: the aux
    target is an (81,) vector over board cells in the same coordinate frame as
    `planes`/`policy`, so it MUST be permuted by the identical symmetry `s`, or
    the label is rotated/reflected relative to its position. Returns 8 triples
    (aug_planes, aug_policy, aug_aux_policy)."""
    return [
        (_sym_board(planes, s), _sym_policy(policy, s), _sym_policy(aux_policy, s))
        for s in range(8)
    ]


def augment_with_cell_targets(
    planes: np.ndarray, policy: np.ndarray, cell_targets: list[np.ndarray]
) -> list[tuple[np.ndarray, np.ndarray, list[np.ndarray]]]:
    """Like `augment`, but ALSO carries an arbitrary list of per-cell (81,)
    target vectors (e.g. the opponent-reply aux target and/or the ownership
    target) through the SAME D4 symmetry as the position.

    Every entry of `cell_targets` is an (81,) vector over board cells in the
    same coordinate frame as `planes`/`policy`, so each MUST be permuted by the
    identical symmetry `s`, or the label is rotated/reflected relative to the
    board it describes (the load-bearing alignment guarantee for the aux heads).
    Returns 8 triples (aug_planes, aug_policy, [aug_target0, aug_target1, ...]).
    """
    return [
        (
            _sym_board(planes, s),
            _sym_policy(policy, s),
            [_sym_policy(t, s) for t in cell_targets],
        )
        for s in range(8)
    ]


def action_to_str(action: int) -> str:
    r, c = divmod(action, BOARD_SIZE)
    return f"{chr(ord('a') + c)}{r + 1}"


def str_to_action(s: str) -> int:
    s = s.strip().lower()
    if len(s) < 2:
        raise ValueError(f"bad move: {s!r}")
    col = ord(s[0]) - ord('a')
    row = int(s[1:]) - 1
    if not (0 <= col < BOARD_SIZE and 0 <= row < BOARD_SIZE):
        raise ValueError(f"out of range: {s!r}")
    return row * BOARD_SIZE + col


def render(state: GameState, *, last_action: int | None = None) -> str:
    """Render board to a string. Shows X for side-to-move's stones, O for opponent."""
    # We render from "the player about to move" perspective:
    # 'X' = side to move's stones, 'O' = opponent's.
    lines = []
    header = "   " + " ".join(chr(ord('a') + c) for c in range(BOARD_SIZE))
    lines.append(header)
    for r in range(BOARD_SIZE):
        row = [f"{r + 1:2d} "]
        for c in range(BOARD_SIZE):
            if state.board[0, r, c]:
                ch = 'X'
            elif state.board[1, r, c]:
                ch = 'O'
            else:
                ch = '.'
            if last_action is not None and last_action == r * BOARD_SIZE + c:
                ch = f"[{ch}]"
                row.append(ch)
            else:
                row.append(f" {ch}")
        lines.append("".join(row))
    return "\n".join(lines)
